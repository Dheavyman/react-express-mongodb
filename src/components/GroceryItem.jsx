import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

const propTypes = {
  item: PropTypes.shape({
    name: PropTypes.string,
    purchased: PropTypes.bool
  }).isRequired,
  actions: PropTypes.shape({
    updateGroceryItem: PropTypes.func.isRequired,
    deleteGroceryItem: PropTypes.func.isRequired
  }).isRequired
};

/**
 * Grocery item component
 *
 * @class GroceryItem
 * @extends {Component}
 */
class GroceryItem extends Component {
  /**
   * Creates an instance of GroceryItem.
   *
   * @param {object} props - Properties passed to component
   * @memberof GroceryItem
   */
  constructor(props) {
    super(props);
    this.togglePurchased = this.togglePurchased.bind(this);
    this.deleteItem = this.deleteItem.bind(this);
  }

  /**
   * Toggle grocery item purchased property
   *
   * @param {object} event - Event
   *
   * @return {any} - None
   * @memberof GroceryItem
   */
  togglePurchased(event) {
    event.preventDefault();
    const { item, actions } = this.props;

    const itemUpdate = {
      ...item,
      purchased: !item.purchased
    }
    actions.updateGroceryItem(itemUpdate)
  }

  /**
   * Delete grocery item
   *
   * @param {object} event - Event
   *
   * @returns {any} - None
   * @memberof GroceryItem
   */
  deleteItem(event) {
    event.preventDefault();
    const { item, actions } = this.props;

    actions.deleteGroceryItem(item)
  }

  /**
   * Render method
   *
   * @returns {object} React element
   * @memberof GroceryItem
   */
  render() {
    const { item } = this.props;

    return (
      <div className="col-sm-12">
        <div className="row">
          <div className="col-5 col-lg-6">
            <h4
              className={classNames('grocery-item', {
                'strike-through': item.purchased
              })}
            >
              {item.name}
            </h4>
          </div>
          <div className="col-3 col-lg-2">
            <form className="three-columns" onSubmit={this.togglePurchased}>
              <button
                type="submit"
                className={classNames('btn', {
                  'btn-primary': !item.purchased,
                  'btn-outline-dark': item.purchased,
                  'btn-return': item.purchased
                })}
              >
                {item.purchased ? 'Return' : 'Buy'}
              </button>
            </form>
          </div>
          <div className="col-4 col-lg-4">
            <form onSubmit={this.deleteItem}>
              <button
                type="submit"
                className="btn btn-outline-danger"
              >
                &times;
              </button>
            </form>
          </div>
        </div>
        <hr />
      </div>
    );
  }
}

GroceryItem.propTypes = propTypes;

export default GroceryItem;
