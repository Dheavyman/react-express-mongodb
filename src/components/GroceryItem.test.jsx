import React from 'react';
import TestRenderer from 'react-test-renderer';

import GroceryItem from './GroceryItem';

describe('GroceryItem component', () => {
  let props;
  beforeEach(() => {
    props = {
      item: {
        name: 'Honey',
        purchased: false
      },
      actions: {
        updateGroceryItem: jest.fn(),
        deleteGroceryItem: jest.fn()
      }
    };
  });
  it('should render the component', () => {
    const renderer = TestRenderer.create(<GroceryItem {...props} />);

    expect(renderer.toJSON()).toMatchSnapshot();
  });
  it('should call togglePurchased handler on click', () => {
    const renderer = TestRenderer.create(<GroceryItem {...props} />);

    renderer.root.findAllByType('form')[0].props.onSubmit(
      { preventDefault: jest.fn()});

    expect(props.actions.updateGroceryItem).toHaveBeenCalled();
  });
  it('should call deleteItem handler on click', () => {
    const renderer = TestRenderer.create(<GroceryItem {...props} />);

    renderer.root.findAllByType('form')[1].props.onSubmit(
      { preventDefault: jest.fn()});

    expect(props.actions.deleteGroceryItem).toHaveBeenCalled();
  });
})
