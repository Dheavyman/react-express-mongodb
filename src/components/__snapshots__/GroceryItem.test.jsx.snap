// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroceryItem component should render the component 1`] = `
<div
  className="col-sm-12"
>
  <div
    className="row"
  >
    <div
      className="col-5 col-lg-6"
    >
      <h4
        className="grocery-item"
      >
        Honey
      </h4>
    </div>
    <div
      className="col-3 col-lg-2"
    >
      <form
        className="three-columns"
        onSubmit={[Function]}
      >
        <button
          className="btn btn-primary"
          type="submit"
        >
          Buy
        </button>
      </form>
    </div>
    <div
      className="col-4 col-lg-4"
    >
      <form
        onSubmit={[Function]}
      >
        <button
          className="btn btn-outline-danger"
          type="submit"
        >
          Ã—
        </button>
      </form>
    </div>
  </div>
  <hr />
</div>
`;
